plugin/evm/limitorders/memory_database.go

```diff
@@ -65,33 +132,113 @@ func (db *InMemoryDatabase) UpdateFilledBaseAssetQuantity(quantity uint, signatu
 	}
 }
 
-// Deletes silently
-func (db *InMemoryDatabase) Delete(signature []byte) {
-	deleteOrder(db, signature)
+func (db *InMemoryDatabase) GetNextFundingTime() uint64 {
+	return db.nextFundingTime
+}
+
+func (db *InMemoryDatabase) UpdateNextFundingTime() {
+	db.nextFundingTime = uint64(getNextHour().Unix())
 }
 
-func (db *InMemoryDatabase) GetLongOrders() []LimitOrder {
+func (db *InMemoryDatabase) GetLongOrders(market Market) []LimitOrder {
 	var longOrders []LimitOrder
 	for _, order := range db.orderMap {
-		if order.PositionType == "long" {
+		if order.PositionType == "long" && order.Market == market {
 			longOrders = append(longOrders, *order)
 		}
 	}
 	sortLongOrders(longOrders)
 	return longOrders
 }
 
-func (db *InMemoryDatabase) GetShortOrders() []LimitOrder {
+func (db *InMemoryDatabase) GetShortOrders(market Market) []LimitOrder {
 	var shortOrders []LimitOrder
 	for _, order := range db.orderMap {
-		if order.PositionType == "short" {
+		if order.PositionType == "short" && order.Market == market {
 			shortOrders = append(shortOrders, *order)
 		}
 	}
 	sortShortOrders(shortOrders)
 	return shortOrders
 }
 
+func (db *InMemoryDatabase) UpdateMargin(trader common.Address, collateral Collateral, addAmount float64) {
+	if _, ok := db.traderMap[trader]; !ok {
+		db.traderMap[trader] = &Trader{
+			Positions: map[Market]Position{},
+			Margins:   map[Collateral]float64{},
+		}
+	}
+
+	if _, ok := db.traderMap[trader].Margins[collateral]; !ok {
+		db.traderMap[trader].Margins[collateral] = 0
+	}
+
+	db.traderMap[trader].Margins[collateral] += addAmount
+}
+
+func (db *InMemoryDatabase) UpdatePosition(trader common.Address, market Market, size float64, openNotional float64) {
+	if _, ok := db.traderMap[trader]; !ok {
+		db.traderMap[trader] = &Trader{
+			Positions: map[Market]Position{},
+			Margins:   map[Collateral]float64{},
+		}
+	}
+
+	if _, ok := db.traderMap[trader].Positions[market]; !ok {
+		db.traderMap[trader].Positions[market] = Position{}
+	}
+
+	position := db.traderMap[trader].Positions[market]
+
+	position.Size = size
+	position.OpenNotional = openNotional
+	db.traderMap[trader].Positions[market] = position
+}
+
+func (db *InMemoryDatabase) UpdateUnrealisedFunding(market Market, fundingRate float64) {
+	for addr, trader := range db.traderMap {
+		position := trader.Positions[market]
+		newFunding := position.Size * fundingRate
+		position.UnrealisedFunding = trader.Positions[market].UnrealisedFunding + newFunding
+		db.traderMap[addr].Positions[market] = position
+	}
+}
+
+func (db *InMemoryDatabase) ResetUnrealisedFunding(market Market, trader common.Address) {
+	if db.traderMap[trader] != nil {
+		if _, ok := db.traderMap[trader].Positions[market]; ok {
+			position := db.traderMap[trader].Positions[market]
```

Comment:
because `db.traderMap[trader].Positions[market].UnrealisedFunding = 0` leads to an error. `db.traderMap[trader].Positions[market].UnrealisedFunding` cannot be assigned directly

https://stackoverflow.com/a/32751792/2485594

One of the comments here says that a pointer to struct can be used instead of a struct for assigning directly but i haven't tried it myself